File: src/day10.py
import dataclasses
from typing import List

import z3


@dataclasses.dataclass
class Switchboard:
    switches: List[List[int]]
    joltages: List[int]

    @staticmethod
    def parse(line: str) -> 'Switchboard':
        sj = line.split("]")[1]
        s, j = sj.split("{")
        switches = [[int(y) for y in x.replace("(", "").replace(")", "").split(",")] for x in s.strip().split(" ")]
        joltages = [int(x) for x in j[:-1].split(",")]
        return Switchboard(switches, joltages)

    def minimum_presses(self):
        how_many_times = [z3.Int(f"s_{i}") for i in range(len(self.switches))]
        solver = z3.Optimize()
        for press in how_many_times:
            solver.add(press >= 0)
        for i, joltage in enumerate(self.joltages):
            res = 0
            for j, switch in enumerate(self.switches):
                if i in switch:
                    res += how_many_times[j]
            solver.add(joltage == res)
        cost = z3.Int('cost')
        solver.add(sum(how_many_times) == cost)
        solver.minimize(cost)
        solver.check()
        return solver.model()[cost].as_long()


def parse(content: str) -> List[Switchboard]:
    return [Switchboard.parse(line) for line in content.split("\n")]


def main():
    content = open("10.txt", "r").read()
    switchboards = parse(content)
    print(sum([switchboard.minimum_presses() for switchboard in switchboards]))


main()

File: src/day10.rs
use itertools::Itertools;
use std::fs;
use std::str::FromStr;

struct Switchboard {
    target: Vec<usize>,
    switches: Vec<Vec<usize>>,
    joltages: Vec<usize>,
}

impl Switchboard {
    fn new(line: &str) -> Self {
        let (t, rest) = line.split(']').collect_tuple().unwrap();
        let target = t
            .chars()
            .skip(1)
            .map(|c| match (c) {
                '.' => 0,
                '#' => 1,
                _ => panic!(),
            })
            .collect();
        let (s, j) = rest.split("{").collect_tuple().unwrap();
        let switches = s
            .trim()
            .split(" ")
            .map(|switch| {
                switch
                    .replace("(", "")
                    .replace(")", "")
                    .split(",")
                    .map(usize::from_str)
                    .map(Result::unwrap)
                    .collect()
            })
            .collect();
        let joltages = j
            .replace("}", "")
            .split(",")
            .map(usize::from_str)
            .map(Result::unwrap)
            .collect();
        Self {
            target,
            switches,
            joltages,
        }
    }

    fn check(&self, switches: &Vec<usize>) -> bool {
        let mut res = vec![0; self.target.len()];
        let selected: Vec<&Vec<usize>> = switches
            .iter()
            .map(|idx| self.switches.get(*idx).unwrap())
            .collect();
        for switch in selected {
            for flip in switch {
                res[*flip] = (res[*flip] + 1) % 2;
            }
        }
        res == self.target
    }

    fn least_presses(&self) -> usize {
        for presses in 1..self.switches.len() {
            for combination in (0..self.switches.len()).into_iter().combinations(presses) {
                if self.check(&combination) {
                    return presses;
                }
            }
        }
        0
    }
}

fn part1(switchboards: &Vec<Switchboard>) -> usize {
    switchboards
        .iter()
        .map(move |switchboard| switchboard.least_presses())
        .sum()
}

pub(crate) fn solve() {
    let contents = fs::read_to_string("10.txt").unwrap();
    let switchboards = contents
        .lines()
        .map(|line| Switchboard::new(line))
        .collect();
    println!("{}", part1(&switchboards));
}
